x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyExecutorTests.java

```diff
@@ -0,0 +1,136 @@
+        ThreadPool.terminate(testThreadPool, 30, TimeUnit.SECONDS);
+        testThreadPool = null;
+    }
+
+    private static class BlockingTestPolicyRunner implements Runnable {
+        private final CountDownLatch latch;
+        private final ActionListener<PolicyExecutionResult> listener;
+
+        BlockingTestPolicyRunner(CountDownLatch latch, ActionListener<PolicyExecutionResult> listener) {
+            this.latch = latch;
+            this.listener = listener;
+        }
+
+        @Override
+        public void run() {
+            try {
+                latch.await();
+                listener.onResponse(new PolicyExecutionResult(true));
+            } catch (InterruptedException e) {
+                throw new RuntimeException("Interrupted waiting for test framework to continue the test", e);
+            }
+        }
+    }
+
+    private static class EnrichPolicyTestExecutor extends EnrichPolicyExecutor {
+
+        EnrichPolicyTestExecutor(Settings settings, ClusterService clusterService, Client client, ThreadPool threadPool,
+                                 IndexNameExpressionResolver indexNameExpressionResolver, LongSupplier nowSupplier) {
+            super(settings, clusterService, client, threadPool, indexNameExpressionResolver, nowSupplier);
+        }
+
+        private CountDownLatch currentLatch;
+        CountDownLatch testRunPolicy(String policyName, EnrichPolicy policy, ActionListener<PolicyExecutionResult> listener) {
+            currentLatch = new CountDownLatch(1);
+            runPolicy(policyName, policy, listener);
+            return currentLatch;
+        }
+
+        @Override
+        protected Runnable createPolicyRunner(String policyName, EnrichPolicy policy, ActionListener<PolicyExecutionResult> listener) {
+            if (currentLatch == null) {
+                throw new IllegalStateException("Use the testRunPolicy method on this test instance");
+            }
+            return new BlockingTestPolicyRunner(currentLatch, listener);
+        }
+    }
+
+    public void testNonConcurrentPolicyExecution() throws InterruptedException {
+        String testPolicyName = "test_policy";
+        EnrichPolicy testPolicy = new EnrichPolicy(EnrichPolicy.EXACT_MATCH_TYPE, null, List.of("some_index"), "keyfield",
+            List.of("valuefield"));
+        EnrichPolicyTestExecutor testExecutor = new EnrichPolicyTestExecutor(Settings.EMPTY, null, null, testThreadPool,
+            new IndexNameExpressionResolver(), ESTestCase::randomNonNegativeLong);
+
+        // Launch a fake policy run that will block until firstTaskBlock is counted down.
+        CountDownLatch firstTaskComplete = new CountDownLatch(1);
+        CountDownLatch firstTaskBlock = testExecutor.testRunPolicy(testPolicyName, testPolicy,
+            new LatchedActionListener<>(noOpListener, firstTaskComplete));
+
+        // Launch a second fake run that should fail immediately because the lock is obtained.
+        Exception expected = null;
```

Comment:
I think you can use expectThrows from LuceneTestCase to help clean this up a bit.

Reply:
Shiny, will do